lib/ex_wire/adapter/tcp.ex

```diff
@@ -0,0 +1,266 @@
+
+        # Let's inform any subscribers
+        if not is_nil(packet) do
+          for {module, function, args} <- Map.get(state, :subscribers, []) do
+            apply(module, function, [packet | args])
+          end
+        end
+
+        updated_state = Map.merge(handled_state, %{secrets: updated_secrets, queued_data: <<>>})
+
+        # If we have more packet data, we need to continue processing.
+        if byte_size(frame_rest) == 0 do
+          {:noreply, updated_state}
+        else
+          handle_info({:tcp, socket, frame_rest}, updated_state)
+        end
+      {:error, "Insufficent data"} ->
+        {:noreply, Map.put(state, :queued_data, total_data)}
+      {:error, reason} ->
+        Logger.error("[Network] Failed to read incoming packet from #{host} `#{reason}`)")
+
+        {:noreply, state}
+    end
+  end
+
+  def handle_info({:tcp_closed, _socket}, state) do
+    Logger.warn("[Network] Peer closed connection.")
+
+    {:noreply, Map.put(state, :active, false)}
+  end
+
+  @doc """
+  If we receive a `send` before secrets are set, we'll send the data directly over the wire.
+  """
+  def handle_cast({:send, %{data: data}}, state = %{socket: socket, host: host}) do
+    Logger.debug("[Network] Sending raw data message of length #{byte_size(data)} byte(s) to #{host}")
+
+    :ok = :gen_tcp.send(socket, data)
+
+    {:noreply, state}
+  end
+
+  @doc """
+  If we receive a `send` and we have secrets set, we'll send the message as a framed Eth packet.
+  """
+  def handle_cast({:send, %{packet: {packet_mod, packet_type, packet_data}}}=data, state = %{host: host, active: false}) when packet_mod != ExWire.Packet.Hello do
+    Logger.info("[Network] Queueing packet #{Atom.to_string(packet_mod)} to #{host}")
+
+    # TODO: Should we monitor this process, etc?
+    pid = self()
+    spawn fn ->
+      :timer.sleep(500)
+      GenServer.cast(pid, data)
+    end
+
+    {:noreply, state}
+  end
+
+  def handle_cast({:send, %{packet: {packet_mod, packet_type, packet_data}}}, state = %{socket: socket, secrets: secrets, host: host}) do
+    Logger.info("[Network] Sending packet #{Atom.to_string(packet_mod)} to #{host}")
+
+    {frame, updated_secrets} = Frame.frame(packet_type, packet_data, secrets)
+
+    :ok = :gen_tcp.send(socket, frame)
+
+    {:noreply, Map.merge(state, %{secrets: updated_secrets})}
+  end
+
+  @doc """
+  Client function for sending a packet over to a peer.
+  """
+  @spec send_packet(pid(), struct()) :: :ok
+  def send_packet(pid, packet) do
+    {:ok, packet_type} = Packet.get_packet_type(packet)
+    {:ok, packet_mod} = Packet.get_packet_mod(packet_type)
+    packet_data = packet_mod.serialize(packet)
+
+    GenServer.cast(pid, {:send, %{packet: {packet_mod, packet_type, packet_data}}})
+
+    :ok
+  end
+
+  @doc """
+  Client function to send HELLO message after connecting.
+  """
+  def send_hello(pid) do
+    send_packet(pid, %Packet.Hello{
+      p2p_version: 0x04,
+      client_id: "Exthereum/0.1",
```

Comment:
Pull this and line 253 into environment variables?

Reply:
Yeah, definitely going to do some minor re-factors on this stuff.